File: MPC/Plant.py
# File: System.py
# Classes: Model()

import numpy as np

def TaylorMethod(F, x, u=[None], dt=1e-3):
    return x + dt*F( x, u );

class Model:
    def __init__(self, F, dt=1e-3,
            model_type='discrete'):
        # default to TaylorMethod when given model is continuous
        if model_type == 'continuous':
            self.model = lambda x,u=[None]: TaylorMethod( F, x, u, dt );
        else:
            self.model = F;

        # set time-step for reference
        self.dt = dt;

    def prop(self, x, u=[None]):
        return self.model( x, u );

File: MPC/cpp/Plant.cpp

// #include <iostream>
#include "Helpers.h"
#include "Plant.h"
// #include <Eigen/Dense>
using Eigen::MatrixXd;

// CONSTRUCTORS:
// Input(s):
//      f: Model function.
// Default variables:
//      time_step = 1e-3
//      model_type = "discrete"
Plant::Plant(MatrixXd (*f)(const MatrixXd &, const MatrixXd &)):
    model(f),
    model_type("discrete"),
    time_step(1e-3),
    mt('d') {}

// Input(s):
//      f: Model function.
//      model_type: Either continuous or discrete.
//      time_step: Length of time-step.
Plant::Plant(MatrixXd (*f)(const MatrixXd &, const MatrixXd &), const std::string &type, const double &dt):
    model(f),
    model_type(type),
    time_step(dt),
    mt(type[0]) {}

// ACCESSOR FUNCTIONS:
// Function: Plant.getModelType()
// Output: Model type (continuous/discrete).
std::string Plant::getModelType()
{
    return model_type;
}

// Function: Plant.getTimeStep()
// Output: Time step of plant model (double).
double Plant::getTimeStep()
{
    return time_step;
}

// MEMBER FUNCTIONS:
// Function: Plant.prop()
// Input(s):
//      x: State of plant at time, t.
//      u: User-input (control term).
// Output(s): Next step of plant simulation.
MatrixXd Plant::prop(const MatrixXd &x, const MatrixXd &u)
{
    if (mt == 'c') {
        return TaylorMethod(model, x, u, time_step);
    }
    return model(x, u);
}
}
